src/coreComponents/physicsSolvers/SolverBase.cpp

```diff
@@ -458,162 +562,223 @@ real64 SolverBase::nonlinearImplicitStep( real64 const & time_n,
-      }
+        }
+        if( m_assemblyCallback )
+        {
+          // Make a copy of LA objects and ship off to the callback
+          array1d< real64 > localRhsCopy( m_rhs.localSize() );
+          localRhsCopy.setValues< parallelDevicePolicy<> >( m_rhs.values() );
+          m_assemblyCallback( m_localMatrix, std::move( localRhsCopy ) );
+        }
 
-      // if the residual norm is less than the Newton tolerance we denote that we have
-      // converged and break from the Newton loop immediately.
+        // TODO: maybe add scale function here?
+        // Scale()
 
-      if( residualNorm < newtonTol && newtonIter >= minNewtonIter )
-      {
-        isConverged = 1;
-        break;
-      }
+        // get residual norm
+        real64 residualNorm = calculateResidualNorm( domain, m_dofManager, m_rhs.values() );
 
-      // do line search in case residual has increased
-      if( m_nonlinearSolverParameters.m_lineSearchAction != NonlinearSolverParameters::LineSearchAction::None
-          && residualNorm > lastResidual )
-      {
-        residualNorm = lastResidual;
-        bool lineSearchSuccess = lineSearch( time_n,
-                                             stepDt,
-                                             cycleNumber,
-                                             domain,
-                                             m_dofManager,
-                                             m_localMatrix.toViewConstSizes(),
-                                             m_rhs,
-                                             m_solution,
-                                             scaleFactor,
-                                             residualNorm );
-
-        if( !lineSearchSuccess )
+
+        GEOSX_LOG_LEVEL_RANK_0( 1, GEOSX_FMT( "    ( R ) = ( {:4.2e} ) ; ", residualNorm ) );
+        if( newtonIter > 0 )
+        {
+          GEOSX_LOG_LEVEL_RANK_0( 1, GEOSX_FMT( "    Last LinSolve(iter,res) = ( {:3}, {:4.2e} ) ; ",
+                                                m_linearSolverResult.numIterations,
+                                                m_linearSolverResult.residualReduction ) );
+        }
+
+        // if the residual norm is less than the Newton tolerance we denote that we have
+        // converged and break from the Newton loop immediately.
+
+        if( residualNorm < newtonTol && newtonIter >= minNewtonIter )
+        {
+          isNewtonConverged = true;
+          break;
+        }
+
+        // do line search in case residual has increased
+        if( m_nonlinearSolverParameters.m_lineSearchAction != NonlinearSolverParameters::LineSearchAction::None
+            && residualNorm > lastResidual )
         {
-          if( m_nonlinearSolverParameters.m_lineSearchAction == NonlinearSolverParameters::LineSearchAction::Attempt )
+          bool lineSearchSuccess = false;
+          if( m_nonlinearSolverParameters.m_lineSearchInterpType == NonlinearSolverParameters::LineSearchInterpolationType::Linear )
```

Comment:
Maybe build a `NonlinearSolverParameters::LineSearchInterpolationType -> callable` map so you do not have to these long lines there?